programs/conditional_vault/src/instructions/redeem_tokens.rs

```diff
@@ -0,0 +1,127 @@
+        let user_underlying_balance_before = accs.user_underlying_token_account.amount;
+        let vault_underlying_balance_before = accs.vault_underlying_token_account.amount;
+        // safe because there is always at least two conditional tokens and thus
+        // at least two user conditional token accounts
+        let max_redeemable = user_conditional_token_accounts
+            .iter()
+            .map(|account| account.amount)
+            .max()
+            .unwrap();
+
+        let mut total_redeemable = 0;
+
+        for (conditional_mint, user_conditional_token_account) in conditional_token_mints
+            .iter()
+            .zip(user_conditional_token_accounts.iter())
+        {
+            // this is safe because we check that every conditional mint is a part of the vault
+            let payout_index = vault
+                .conditional_token_mints
+                .iter()
+                .position(|mint| mint == &conditional_mint.key())
+                .unwrap();
+
+            total_redeemable += ((user_conditional_token_account.amount as u128
+                * question.payout_numerators[payout_index] as u128)
+                / question.payout_denominator as u128) as u64;
+
+            token::burn(
+                CpiContext::new(
+                    accs.token_program.to_account_info(),
+                    Burn {
+                        mint: conditional_mint.to_account_info(),
+                        from: user_conditional_token_account.to_account_info(),
+                        authority: accs.authority.to_account_info(),
+                    },
+                ),
+                user_conditional_token_account.amount,
+            )?;
+        }
+
+        token::transfer(
+            CpiContext::new_with_signer(
+                accs.token_program.to_account_info(),
+                Transfer {
+                    from: accs.vault_underlying_token_account.to_account_info(),
+                    to: accs.user_underlying_token_account.to_account_info(),
+                    authority: accs.vault.to_account_info(),
+                },
+                signer,
+            ),
+            total_redeemable,
+        )?;
+
+        require_gte!(max_redeemable, total_redeemable, VaultError::AssertFailed);
+
+        ctx.accounts.user_underlying_token_account.reload()?;
+        ctx.accounts.vault_underlying_token_account.reload()?;
+
+        require_eq!(
+            ctx.accounts.user_underlying_token_account.amount,
+                user_underlying_balance_before + total_redeemable,
+                VaultError::AssertFailed
+        );
+
+        require_eq!(
+            ctx.accounts.vault_underlying_token_account.amount,
+                vault_underlying_balance_before - total_redeemable,
+                VaultError::AssertFailed
+        );
+
+        for acc in user_conditional_token_accounts.iter_mut() {
+            acc.reload()?;
+            require_eq!(acc.amount, 0, VaultError::AssertFailed);
```

Comment:
might as well close the accounts

Reply:
makes it easier for me to index stuff if they're not closed